### ISSUE TITLE
Compiler error invoked with gcc 6.1.1

### ISSUE THREAD

**Issue by jaked122** [CleverRaven/Cataclysm-DDA#16725]
When using the newest release of gcc(6.1.1) to compile Catadda, it appears that it has issues dealing with certain parts of the source, namely the sdl version's tile code.

> src/cata_tiles.cpp: In member function ‘void cata_tiles::draw(int, int, const tripoint&, int, int)’:
> src/cata_tiles.cpp:972:54: internal compiler error: in add_expr, at tree.c:7870
>                          &cata_tiles::draw_critter_at } ) {

Now naturally, this isn't really a problem for catadda, but I'm not exactly sure about how to go about reporting it, as this is a very large project.

I'm having trouble getting g++ to generate the correct set of preprocessed files to send to them.

Thanks for any help.

**Comment by franz-mokka** [comment]
Also happening to me on Arch with GCC 6.1.1.

To generate the preprocessed files, I believe g++ can do it with the `-E` option, so maybe you can do add it somewhere in the Makefile or run make as `CXXFLAGS=-E make TILES=1`.

From the C:DDA side, this error can be avoided by taking the drawing layers array out of the for loop in cata_tiles.cpp:

```diff
diff --git a/src/cata_tiles.cpp b/src/cata_tiles.cpp
index acca8f4..580a912 100644
--- a/src/cata_tiles.cpp
+++ b/src/cata_tiles.cpp
@@ -965,11 +965,19 @@ void cata_tiles::draw( int destx, int desty, const tripoint &center, int width,
 
             draw_points.push_back( tile_render_info( tripoint( x, y, center.z ), height_3d ) );
         }
+
+        const decltype( &cata_tiles::draw_furniture ) drawing_layers[] = {
+            &cata_tiles::draw_furniture,
+            &cata_tiles::draw_trap,
+            &cata_tiles::draw_field_or_item,
+            &cata_tiles::draw_vpart,
+            &cata_tiles::draw_vpart_below,
+            &cata_tiles::draw_terrain_below,
+            &cata_tiles::draw_critter_at
+        };
+
         // for each of the drawing layers in order, back to front ...
-        for( auto f : { &cata_tiles::draw_furniture, &cata_tiles::draw_trap,
-                        &cata_tiles::draw_field_or_item, &cata_tiles::draw_vpart,
-                        &cata_tiles::draw_vpart_below, &cata_tiles::draw_terrain_below,
-                        &cata_tiles::draw_critter_at } ) {
+        for( auto f : drawing_layers ) {
             // ... draw all the points we drew terrain for, in the same order
             for( auto &p : draw_points ) {
                 (this->*f)( p.pos, ch.visibility_cache[p.pos.x][p.pos.y], p.height_3d );
```

**Comment by jaked122** [comment]
I just reported it on the gcc bugzilla. Hopefully they'll take notice relatively soon.

I'm going to create a PR though for this, if you don't mind.

**Comment by BrettDong** [comment]
@illi-kun This has been fixed in #16757 though it is not merged on GitHub but directly picked into Git by Kevin Granade.

**Comment by illi-kun** [comment]
@BrettDong thanks for notice.
